sys/parsoid_proxy.js

```diff
@@ -0,0 +1,161 @@
+"use strict";
+
+const P = require('bluebird');
+const HyperSwitch = require('hyperswitch');
+const URI = HyperSwitch.URI;
+
+const mwUtil = require('../lib/mwUtil');
+
+const spec = HyperSwitch.utils.loadSpec(`${__dirname}/parsoid.yaml`);
+
+class ParsoidProxy {
+    constructor(options) {
+        this.options = options || {};
+        this.options.backends = this.options.backends || { default: 'old' };
+
+        // Set up operations
+        this.operations = {
+            getPageBundle: this.pagebundle.bind(this),
+            // Revision retrieval per format
+            getWikitext: this.getFormat.bind(this, 'wikitext'),
+            getHtml: this.getFormat.bind(this, 'html'),
+            getDataParsoid: this.getFormat.bind(this, 'data-parsoid'),
+            // Listings
+            listWikitextRevisions: this.listRevisions.bind(this, 'wikitext'),
+            listHtmlRevisions: this.listRevisions.bind(this, 'html'),
+            listDataParsoidRevisions: this.listRevisions.bind(this, 'data-parsoid'),
+            // Transforms
+            transformHtmlToHtml: this.makeTransform('html', 'html'),
+            transformHtmlToWikitext: this.makeTransform('html', 'wikitext'),
+            transformWikitextToHtml: this.makeTransform('wikitext', 'html'),
+            transformWikitextToLint: this.makeTransform('wikitext', 'lint'),
+            transformChangesToWikitext: this.makeTransform('changes', 'wikitext')
+        };
+    }
+
+    _chooseBackend(req) {
+        return this.options.backends[req.params.domain] || this.options.backends.default;
+    }
+
+    _buildRequest(prefix, req, listing) {
+        const rp = req.params;
+        const path = [rp.domain, 'sys'].concat(prefix);
+        if (rp.title) {
+            path.push(rp.title);
+            if (rp.revision) {
+                path.push(rp.revision);
+                if (rp.tid) {
+                    path.push(rp.tid);
+                }
+            }
+        }
+        if (listing) {
+            path.push('');
+        }
+        return {
+            method: req.method,
+            uri: new URI(path),
+            query: req.query,
+            headers: req.headers,
+            body: req.body
+        };
+    }
+
+    _createDoubleProcessingRequest(hyper, req, oldPath, newPath, listing) {
```

Comment:
Since `parsoid_old` and `parsoid_new` never change, perhaps it would be enough to have just one `path` param and prefix it with the appropriate component directly here below while making the requests?